pkg/quota/controller/clusterquotamapping/clusterquotamapping.go

```diff
@@ -0,0 +1,585 @@
+// that it would be nearly impossible for any resources to be created.  If you do create resources, then we must be observing
+// their deletes.  When quota is replenished, we'll see that we need to clear any charges.
+//
+// When going from namespace to quota, you can get back a quota that doesn't exist.  Since the cache is shared,
+// we know that a missing quota means that there isn't anything for us to bill against, so we can skip it.
+//
+// If the mapping cache is wrong and a previously deleted quota or namespace is created, this controller
+// correctly adds the items back to the list and clears out all previous mappings.
+//
+// In addition to those constraints, the timing threshold for actaully hitting this problem is really tight.  It's
+// basically a script that is creating and deleting things as fast as it possibly can.  Sub-millisecond in the fuzz
+// test where I caught the problem.
+
+// NewClusterQuotaMappingController builds a mapping between namespaces and clusterresourcequotas
+func NewClusterQuotaMappingController(namespaceInformer shared.NamespaceInformer, quotaInformer shared.ClusterResourceQuotaInformer) *ClusterQuotaMappingController {
+	c := &ClusterQuotaMappingController{
+		namespaceQueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		quotaQueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		clusterQuotaMapper: &clusterQuotaMapper{
+			requiredQuotaToSelector:    map[string]*unversioned.LabelSelector{},
+			requiredNamespaceToLabels:  map[string]map[string]string{},
+			completedQuotaToSelector:   map[string]*unversioned.LabelSelector{},
+			completedNamespaceToLabels: map[string]map[string]string{},
+
+			quotaToNamespaces: map[string]sets.String{},
+			namespaceToQuota:  map[string]sets.String{},
+		},
+	}
+
+	namespaceInformer.Informer().AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addNamespace,
+		UpdateFunc: c.updateNamespace,
+		DeleteFunc: c.deleteNamespace,
+	})
+	c.namespaceLister = namespaceInformer.Lister()
+	c.namespacesSynced = namespaceInformer.Informer().HasSynced
+
+	quotaInformer.Informer().AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addQuota,
+		UpdateFunc: c.updateQuota,
+		DeleteFunc: c.deleteQuota,
+	})
+	c.quotaLister = quotaInformer.Lister()
+	c.quotasSynced = quotaInformer.Informer().HasSynced
+
+	return c
+}
+
+type ClusterQuotaMappingController struct {
+	namespaceQueue   workqueue.RateLimitingInterface
+	namespaceLister  *ocache.IndexerToNamespaceLister
+	namespacesSynced func() bool
+
+	quotaQueue   workqueue.RateLimitingInterface
+	quotaLister  *ocache.IndexerToClusterResourceQuotaLister
+	quotasSynced func() bool
+
+	clusterQuotaMapper *clusterQuotaMapper
+}
+
+type clusterQuotaMapper struct {
+	lock sync.RWMutex
+
+	requiredQuotaToSelector    map[string]*unversioned.LabelSelector
```

Comment:
Oh I see, complete are synced, required need syncing. Can you document this somewhere?